### PR TITLE
HARMONY-1426: Make changes to improve performance to scale up more pods

### PR DESCRIPTION
**Pull request by chris-durbin**
## Jira Issue ID
HARMONY-1426

## Description
After performing testing with a workload run and with a single large 110K granule request we identified some changes to improve throughput and also help ensure synchronous requests can complete in reasonable times while large requests are also running.

## Local Test Steps
You will need to change your .env file for running harmony to change the names of the queues for work item updates since they are now FIFO:
```
WORK_ITEM_UPDATE_QUEUE_URL=http://localhost:4566/queue/work-item-update-queue.fifo
LARGE_WORK_ITEM_UPDATE_QUEUE_URL=http://localhost:4566/queue/large-work-item-update-queue.fifo
```
1. Delete your localstack pod: `kubectl delete pod -n harmony localstack-pod`
2. `bin/start-postgres-localstack` (this will cause the FIFO queues to be created for work item updates)
3. Build the scheduler (I think I changed some scheduler code) `npm run build` in the `kubernetes-services/work-scheduler/` directory.
4. `bin/restart-services`

Make sure requests complete successfully for you.

In sandbox I was able to run with 200 pods and the 110K granule request completed in under four hours. Queues never backed up to more than 60 items for the work item updates and while there were some slower DB queries in locking the jobs table with these changes and the 200 pod concurrency the locking was in the 500-700ms range as opposed to 2 minutes prior to the changes.

## PR Acceptance Checklist
* [X] Acceptance criteria met
* [X] Tests added/updated (if needed) and passing
* [ ] Documentation updated (if needed)

### CODE REVIEW

File: app/backends/workflow-orchestration/work-item-polling.ts [hunk 1]
```diff
@@ -135,8 +134,7 @@ eventEmitter.on(WorkItemEvent.CREATED, async (workItem: WorkItem) => {
     const { serviceID } = workItem;
     defaultLogger.debug(`Work item created for service ${serviceID}, putting message on scheduler queue`);
     const queue = getWorkSchedulerQueue();
-    // must include groupId for FIFO queues, but we don't care about it so just use 'w'
-    await queue.sendMessage(serviceID, 'w');
+    await queue.sendMessage(serviceID);
```
Comment (chris-durbin) [inline app/backends/workflow-orchestration/work-item-polling.ts:137]: Scheduler queue is a regular queue - not FIFO - so no need to pass a group ID.

File: app/backends/workflow-orchestration/workflow-orchestration.ts [hunk 2]
```diff
@@ -97,13 +97,13 @@ export async function updateWorkItem(req: HarmonyRequest, res: Response): Promis
...
-    workItemLogger.debug('Sending work item update to large item queue');
+    workItemLogger.debug(`Sending work item update to large item queue for ${operation.requestId}`);
     queueType = WorkItemQueueType.LARGE_ITEM_UPDATE;
   } else {
-    workItemLogger.debug('Sending work item update to regular queue');
+    workItemLogger.debug(`Sending work item update to regular queue for ${operation.requestId}`);
   }
   const queue = getQueueForType(queueType);
-  await queue.sendMessage(JSON.stringify({ update, operation })).catch((e) => {
+  await queue.sendMessage(JSON.stringify({ update, operation }), operation.requestId).catch((e) => {
```
Comment (chris-durbin) [inline app/backends/workflow-orchestration/workflow-orchestration.ts:106]: One of the more important changes - the work item update queues are now FIFO, and we pass the job ID in as the group ID. This allows us to process the work item updates for a single job serially significantly reducing locking while also allowing other jobs to have their work item updates processed by other processors and not get stuck behind a possible backlog caused by another job.

File: app/models/job.ts [hunk 3]
```diff
@@ -576,7 +576,7 @@ export class Job extends DBRecord implements JobRecord {
     currentPage = 0,
     perPage = env.defaultResultPageSize,
   ): Promise<{ job: Job; pagination: ILengthAwarePagination }> {
-    const result = await transaction('jobs').select().where({ username, requestId }).forUpdate();
+    const result = await transaction('jobs').select().where({ username, requestId });
```
Comment (chris-durbin) [inline app/models/job.ts:579]: The two changes in this file were places where we did not need to lock the jobs table as far as I could tell - they were only used for read only things like displaying job status.

File: app/workers/work-failer.ts [hunk 4]
```diff
@@ -51,7 +51,7 @@ export default class WorkFailer implements Worker {
     const jobServiceThresholds = {};
 
     const workItems = await getWorkItemsByUpdateAgeAndStatus(
-      db, lastUpdateOlderThanMinutes, [WorkItemStatus.RUNNING],
+      db, lastUpdateOlderThanMinutes, [WorkItemStatus.RUNNING, WorkItemStatus.QUEUED],
```
Comment (chris-durbin) [inline app/workers/work-failer.ts:54]: The work failer will now fail (cause a retry) for work items that have been in the 'queued' state for a long time, just like we do with the 'running' state.

File: env-defaults [hunk 5]
```diff
@@ -184,7 +184,7 @@ WORK_FAILER_PERIOD_SEC=120
 # Used to avoid a tight loop
 WORK_ITEM_UPDATE_QUEUE_PROCESSOR_DELAY_AFTER_ERROR_SEC=1
 
-WORK_ITEM_UPDATE_QUEUE_PROCESSOR_COUNT=6
+WORK_ITEM_UPDATE_QUEUE_PROCESSOR_COUNT=3
```
Comment (chris-durbin) [inline env-defaults:187]: All of my successful testing in sandbox was with 3 processors, so changing the default.

**Review by vinnyinverso — APPROVED** [review]

File: kubernetes-services/work-scheduler/env-defaults [hunk 6]
```diff
@@ -34,7 +34,7 @@ QUEUE_LONG_POLLING_WAIT_TIME_SEC=20
 CMR_MAX_PAGE_SIZE=2000
 
 # SQS queue used to request scheduling of work items for a service
-WORK_ITEM_SCHEDULER_QUEUE_URL=http://localstack:4566/queue/work-item-scheduler-queue
+WORK_ITEM_SCHEDULER_QUEUE_URL=http://localstack:4566/queue/work-item-scheduler-queue.fifo
```
Comment (indiejames) [inline kubernetes-services/work-scheduler/env-defaults:37]: Wait - I thought the scheduler queue was _not_ FIFO?

**Review by indiejames — COMMENTED** [review]
Just a small comment about naming the scheduler queue.

File: env-defaults (hunk 5 above)
Comment (chris-durbin) [reply env-defaults:187]: We use the env-defaults in deployed environments as well unless there is a Bamboo variable overriding it.

File: kubernetes-services/work-scheduler/env-defaults (hunk 6 above)
Comment (chris-durbin) [reply kubernetes-services/work-scheduler/env-defaults:37]: oops forgot to revert - good catch

**Review by indiejames — APPROVED** [review]

**Review by ygliuvt — APPROVED** [review]

**Pull request merged**